src/mito/src/engine/procedure/alter.rs

```diff
@@ -0,0 +1,540 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use common_procedure::error::{Error, FromJsonSnafu, ToJsonSnafu};
+use common_procedure::{Context, LockKey, Procedure, ProcedureManager, Result, Status};
+use common_telemetry::logging;
+use serde::{Deserialize, Serialize};
+use snafu::{ensure, OptionExt, ResultExt};
+use store_api::manifest::Manifest;
+use store_api::storage::{AlterRequest, Region, RegionMeta, StorageEngine};
+use table::engine::TableReference;
+use table::metadata::{RawTableInfo, TableInfo, TableVersion};
+use table::requests::{AlterKind, AlterTableRequest};
+use table::Table;
+
+use crate::engine::MitoEngineInner;
+use crate::error::{
+    BuildTableMetaSnafu, TableNotFoundSnafu, UpdateTableManifestSnafu, VersionChangedSnafu,
+};
+use crate::manifest::action::{TableChange, TableMetaAction, TableMetaActionList};
+use crate::table::{create_alter_operation, MitoTable};
+
+/// Procedure to alter a [MitoTable].
+pub(crate) struct AlterMitoTable<S: StorageEngine> {
+    data: AlterTableData,
+    engine_inner: Arc<MitoEngineInner<S>>,
+    table: Arc<MitoTable<S::Region>>,
+    /// The table info after alteration.
+    new_info: Option<TableInfo>,
+}
+
+#[async_trait]
+impl<S: StorageEngine> Procedure for AlterMitoTable<S> {
+    fn type_name(&self) -> &str {
+        Self::TYPE_NAME
+    }
+
+    async fn execute(&mut self, _ctx: &Context) -> Result<Status> {
+        match self.data.state {
+            AlterTableState::Prepare => self.on_prepare(),
+            AlterTableState::AlterRegions => self.on_alter_regions().await,
+            AlterTableState::UpdateTableManifest => self.on_update_table_manifest().await,
+        }
+    }
+
+    fn dump(&self) -> Result<String> {
+        let json = serde_json::to_string(&self.data).context(ToJsonSnafu)?;
+        Ok(json)
+    }
+
+    fn lock_key(&self) -> LockKey {
+        let table_ref = self.data.table_ref();
+        let info = self.table.table_info();
+        let mut keys: Vec<_> = info
+            .meta
+            .region_numbers
+            .iter()
+            .map(|number| format!("{table_ref}/region-{number}"))
+            .collect();
+        // If alter kind is rename, we also need to lock the region with another name.
+        if let AlterKind::RenameTable { new_table_name } = &self.data.request.alter_kind {
+            let new_table_ref = TableReference {
+                catalog: &self.data.request.catalog_name,
+                schema: &self.data.request.schema_name,
+                table: new_table_name,
+            };
+            // We only acquire the first region.
+            keys.push(format!("{new_table_ref}/region-0"));
```

Comment:
Why not use `common_catalog::format_full_table_name`? The `new_table_ref` is unnecessary.

Reply:
format_full_table_name requires additional allocation and other lock keys also use TableReference to format itself.